src/GraphQL/Internal/Validation.hs

```diff
@@ -183,18 +205,42 @@ validateArguments args = mapErrors DuplicateArgument (makeMap [(name, value) | A
 -- imply that everything is valid.
 
 -- | A GraphQL selection.
-data Selection spread
-  = SelectionField (Field spread)
+data Selection' spread
+  = SelectionField (Field' spread)
   | SelectionFragmentSpread spread
   | SelectionInlineFragment (InlineFragment spread)
   deriving (Eq, Show)
 
+-- | Get all of the fields directly inside the given selection set.
+--
+-- TODO: This ignores fragments, whereas it should actually do something with
+-- them.
+--
+-- TODO: At this point, we ought to know that field names are unique. As such,
+-- we should return an ordered map of Name to Fields, rather than a bland
+-- list.
+getFields :: SelectionSet -> [Field]
+getFields ss = [field | SelectionField field <- ss]
+
 -- | A field in a selection set, which itself might have children which might
 -- have fragment spreads.
-data Field spread
-  = Field (Maybe Alias) Name Arguments Directives [Selection spread]
+data Field' spread
+  = Field' (Maybe Alias) Name Arguments Directives [Selection' spread]
   deriving (Eq, Show)
 
+instance HasName (Field' spread) where
+  getName (Field' _ name _ _ _) = name
+
+type Field = Field' FragmentSpread
```

Comment:
Suggestions for better names welcome. The key thing is that it's the type of the fragment spread used by the field, and that the reason these types vary (i.e. why they are type _vari_ables) is that we've got some state machines and this variable represents the current state.